backend/src/worker.rs

```diff
@@ -1665,6 +1662,274 @@ def main():
         assert_eq!(json!({ "l": [0, 1, 2] }), result);
     }
 
+    mod retry {
+        use super::*;
+
+        /// test helper provides some external state to help steps fail at specific points
+        struct Server {
+            addr: std::net::SocketAddr,
+            tx: tokio::sync::oneshot::Sender<()>,
+            task: tokio::task::JoinHandle<Vec<u8>>,
+        }
+
+        impl Server {
+            async fn start(responses: Vec<Option<u8>>) -> Self {
+                use tokio::net::TcpListener;
+
+                let (tx, rx) = tokio::sync::oneshot::channel::<()>();
+                let sock = TcpListener::bind("127.0.0.1:0").await.unwrap();
+                let addr = sock.local_addr().unwrap();
+
+                let task = tokio::task::spawn(async move {
+                    tokio::pin!(rx);
+                    let mut results = vec![];
+
+                    for next in responses {
+                        let (mut peer, _) = tokio::select! {
+                            _ = &mut rx => break,
+                            r = sock.accept() => r,
+                        }
+                        .unwrap();
+
+                        let n = peer.read_u8().await.unwrap();
+                        results.push(n);
+
+                        if let Some(next) = next {
+                            peer.write_u8(next).await.unwrap();
+                        }
+                    }
+
+                    results
+                });
+
+                return Self { addr, tx, task };
+            }
+
+            async fn close(self) -> Vec<u8> {
+                let Self { task, tx, .. } = self;
+                drop(tx);
+                task.await.unwrap()
+            }
+        }
+
+        fn flow() -> FlowValue {
+            let inner = r#"
+            export async function main(index, port) {
+                const buf = new Uint8Array([0]);
+                const sock = await Deno.connect({ port });
+                await sock.write(new Uint8Array([index]));
+                if (1 != await sock.read(buf)) throw "read";
```

Comment:
very nit but: `1 != await sock.read(buf)` seems harder to read than `await sock.read(buf) != 1`